File: backend/app/db/session.py
"""MongoDB client, index setup, and the Unit of Work every repository runs inside.

The Unit of Work is deliberately small: an identity map plus a dirty check. Entities
are tracked when they are added or read, `flush()` writes back only the ones whose
document actually changed, and `commit()` ends the request. When Mongo is a replica
set (`MONGO_TRANSACTIONS=true`) the whole unit runs inside a real multi-document
transaction, so a stage that fails halfway leaves nothing behind; against a standalone
mongod there is no transaction to join, and `rollback()` can only drop writes that have
not been flushed yet. That difference is a property of the deployment, so it is
configuration rather than a code path chosen at import time.
"""
from __future__ import annotations

from collections.abc import Iterator
from contextlib import contextmanager
from functools import lru_cache
from typing import Any

from pymongo import ASCENDING, MongoClient
from pymongo.database import Database

from app.core.config import settings
from app.core.logging import get_logger
from app.db.base import Entity

log = get_logger(__name__)

# (collection, keys, unique) - the uniqueness rules the data model promises, plus the
# org-scoped lookups every repository read performs.
INDEXES: list[tuple[str, list[tuple[str, int]], bool]] = [
    ("users", [("email", ASCENDING)], False),
    ("users", [("organization_id", ASCENDING), ("email", ASCENDING)], True),
    ("assessments", [("organization_id", ASCENDING)], False),
    ("documents", [("organization_id", ASCENDING), ("assessment_id", ASCENDING)], False),
    (
        "documents",
        [
            ("organization_id", ASCENDING),
            ("assessment_id", ASCENDING),
            ("kind", ASCENDING),
            ("checksum", ASCENDING),
        ],
        True,
    ),
    ("pages", [("organization_id", ASCENDING), ("document_id", ASCENDING)], False),
    ("pages", [("document_id", ASCENDING), ("page_number", ASCENDING)], True),
    ("blocks", [("organization_id", ASCENDING), ("page_id", ASCENDING)], False),
    ("questions", [("organization_id", ASCENDING), ("assessment_id", ASCENDING)], False),
    ("question_regions", [("organization_id", ASCENDING), ("question_id", ASCENDING)], False),
    ("answers", [("organization_id", ASCENDING), ("assessment_id", ASCENDING)], False),
    ("answer_regions", [("organization_id", ASCENDING), ("answer_id", ASCENDING)], False),
    (
        "mappings",
        [
            ("organization_id", ASCENDING),
            ("assessment_id", ASCENDING),
            ("review_status", ASCENDING),
        ],
        False,
    ),
    ("grades", [("organization_id", ASCENDING), ("mapping_id", ASCENDING)], False),
    ("jobs", [("organization_id", ASCENDING), ("assessment_id", ASCENDING)], False),
]


@lru_cache
def get_client() -> MongoClient:
    return MongoClient(settings.mongo_uri, uuidRepresentation="standard", tz_aware=True)


def get_database() -> Database:
    return get_client()[settings.mongo_db_name]


def create_all() -> None:
    """Mongo has no schema to migrate; indexes are the only thing to declare."""
    ensure_indexes(get_database())


def ensure_indexes(database: Database) -> None:
    for collection, keys, unique in INDEXES:
        database[collection].create_index(keys, unique=unique)


class UnitOfWork:
    """Identity map + dirty check over a MongoDB database."""

    def __init__(self, database: Database, *, use_transaction: bool | None = None) -> None:
        self.db = database
        self._use_transaction = (
            settings.mongo_transactions if use_transaction is None else use_transaction
        )
        self._client_session: Any | None = None
        self._identity: dict[tuple[str, str], Entity] = {}
        self._snapshots: dict[tuple[str, str], dict[str, Any]] = {}
        self._deleted: dict[tuple[str, str], Entity] = {}

    # ------------------------------------------------------------------- transactions
    @property
    def client_session(self) -> Any | None:
        """The pymongo session to pass to every driver call, or None when transactions
        are off. Repositories pass it so their reads see this unit own writes."""
        if not self._use_transaction:
            return None
        if self._client_session is None:
            self._client_session = self.db.client.start_session()
            self._client_session.start_transaction()
        return self._client_session

    # ------------------------------------------------------------------- identity map
    @staticmethod
    def _key(entity: Entity) -> tuple[str, str]:
        return (type(entity).__collection__, entity.id)

    def track(self, entity: Entity) -> Entity:
        """Register an entity read from Mongo, returning the instance already in the
        map if there is one - two reads of the same row must not diverge in memory."""
        key = self._key(entity)
        if key in self._identity:
            return self._identity[key]
        self._identity[key] = entity
        self._snapshots[key] = entity.to_mongo()
        return entity

    def tracked(self, collection: str, entity_id: str) -> Entity | None:
        return self._identity.get((collection, entity_id))

    # -------------------------------------------------------------------------- writes
    def add(self, entity: Entity) -> Entity:
        if not type(entity).__collection__:
            raise AssertionError(f"{type(entity).__name__} declares no __collection__")
        key = self._key(entity)
        self._identity[key] = entity
        self._snapshots.pop(key, None)  # no snapshot -> treated as an insert on flush
        self._deleted.pop(key, None)
        return entity

    def delete(self, entity: Entity) -> None:
        key = self._key(entity)
        self._identity.pop(key, None)
        self._snapshots.pop(key, None)
        self._deleted[key] = entity

    def flush(self) -> None:
        """Write pending deletes and every entity whose document changed."""
        for (collection, entity_id), _entity in list(self._deleted.items()):
            self.db[collection].delete_one({"_id": entity_id}, session=self.client_session)
        self._deleted.clear()

        for key, entity in list(self._identity.items()):
            collection, entity_id = key
            document = entity.to_mongo()
            snapshot = self._snapshots.get(key)
            if snapshot == document:
                continue
            if snapshot is not None:
                entity.touch()
                document = entity.to_mongo()
            self.db[collection].replace_one(
                {"_id": entity_id}, document, upsert=True, session=self.client_session
            )
            self._snapshots[key] = document

    def commit(self) -> None:
        self.flush()
        if self._client_session is not None:
            self._client_session.commit_transaction()
            self._client_session.end_session()
            self._client_session = None

    def rollback(self) -> None:
        if self._client_session is not None:
            self._client_session.abort_transaction()
            self._client_session.end_session()
            self._client_session = None
        else:
            log.warning("rollback without a transaction: only unflushed writes are discarded")
        self.expire_all()

    def expire_all(self) -> None:
        """Forget every in-memory entity so the next read comes from Mongo."""
        self._identity.clear()
        self._snapshots.clear()
        self._deleted.clear()

    def close(self) -> None:
        if self._client_session is not None:
            self._client_session.abort_transaction()
            self._client_session.end_session()
            self._client_session = None
        self.expire_all()


@contextmanager
def session_scope() -> Iterator[UnitOfWork]:
    unit = UnitOfWork(get_database())
    try:
        yield unit
        unit.commit()
    except Exception:
        unit.rollback()
        raise
    finally:
        unit.close()


def get_session() -> Iterator[UnitOfWork]:
    """FastAPI dependency."""
    unit = UnitOfWork(get_database())
    try:
        yield unit
    finally:
        unit.close()

File: apps/web/lib/server/db/session.ts
/**
 * MongoDB client, index setup, and the Unit of Work every repository runs inside.
 * Port of backend/app/db/session.py.
 *
 * Atlas is always a replica set (even the free M0 tier), so — unlike the Python
 * version, which had to degrade gracefully against a standalone local mongod —
 * this always runs inside a real multi-document transaction. That is a genuine
 * simplification, not just a port: there is no MONGO_TRANSACTIONS flag here.
 *
 * The client is cached on the Node global so warm serverless invocations reuse the
 * same connection pool instead of opening a new one per request (the standard
 * pattern for MongoDB + Vercel/Next.js).
 */
import { MongoClient, type Db } from "mongodb";
import { settings } from "@/lib/server/config";
import { newId, type Entity } from "./base";

const INDEXES: Array<{ collection: string; keys: Record<string, 1>; unique: boolean }> = [
  { collection: "users", keys: { email: 1 }, unique: false },
  { collection: "users", keys: { organizationId: 1, email: 1 }, unique: true },
  { collection: "assessments", keys: { organizationId: 1 }, unique: false },
  { collection: "documents", keys: { organizationId: 1, assessmentId: 1 }, unique: false },
  {
    collection: "documents",
    keys: { organizationId: 1, assessmentId: 1, kind: 1, checksum: 1 },
    unique: true,
  },
  { collection: "pages", keys: { organizationId: 1, documentId: 1 }, unique: false },
  { collection: "pages", keys: { documentId: 1, pageNumber: 1 }, unique: true },
  { collection: "blocks", keys: { organizationId: 1, pageId: 1 }, unique: false },
  { collection: "questions", keys: { organizationId: 1, assessmentId: 1 }, unique: false },
  { collection: "question_regions", keys: { organizationId: 1, questionId: 1 }, unique: false },
  { collection: "answers", keys: { organizationId: 1, assessmentId: 1 }, unique: false },
  { collection: "answer_regions", keys: { organizationId: 1, answerId: 1 }, unique: false },
  {
    collection: "mappings",
    keys: { organizationId: 1, assessmentId: 1, reviewStatus: 1 },
    unique: false,
  },
  { collection: "grades", keys: { organizationId: 1, mappingId: 1 }, unique: false },
  { collection: "jobs", keys: { organizationId: 1, assessmentId: 1 }, unique: false },
];

declare global {
  // eslint-disable-next-line no-var
  var __mongoClientPromise: Promise<MongoClient> | undefined;
}

function getClientPromise(): Promise<MongoClient> {
  if (!global.__mongoClientPromise) {
    const client = new MongoClient(settings.mongoUri, { maxPoolSize: 10 });
    const connecting = client.connect();
    // A rejected promise is still a truthy cached value — without this, one failed
    // connection attempt (a transient network blip, Atlas hiccup, etc.) would wedge
    // every request for the rest of the process's life behind the same permanently-
    // rejected promise, since nothing would ever try `new MongoClient(...)` again.
    connecting.catch(() => {
      if (global.__mongoClientPromise === connecting) global.__mongoClientPromise = undefined;
    });
    global.__mongoClientPromise = connecting;
  }
  return global.__mongoClientPromise;
}

export async function getDatabase(): Promise<Db> {
  const client = await getClientPromise();
  return client.db(settings.mongoDbName);
}

export async function ensureIndexes(): Promise<void> {
  const db = await getDatabase();
  const byCollection = new Map<string, typeof INDEXES>();
  for (const spec of INDEXES) {
    const list = byCollection.get(spec.collection) ?? [];
    list.push(spec);
    byCollection.set(spec.collection, list);
  }

  for (const [collection, specs] of byCollection) {
    const coll = db.collection(collection);
    const declaredNames = new Set(specs.map(({ keys }) => indexName(keys)));
    for (const spec of specs) {
      await coll.createIndex(spec.keys, { unique: spec.unique });
    }
    // Drop anything left over from an earlier schema generation (e.g. a prior
    // snake_case naming convention) that isn't declared above — a stale unique
    // index on fields the current models don't populate silently 11000s every
    // insert past the first, since Mongo indexes the missing fields as null.
    const existing = await coll.listIndexes().toArray();
    for (const index of existing) {
      if (index.name === "_id_" || declaredNames.has(index.name!)) continue;
      await coll.dropIndex(index.name!);
    }
  }
}

function indexName(keys: Record<string, 1>): string {
  return Object.entries(keys)
    .map(([field, dir]) => `${field}_${dir}`)
    .join("_");
}

/**
 * Identity map + dirty check over a MongoDB database.
 *
 * Writes are flushed inside a short-lived multi-document transaction (Atlas is
 * always a replica set) so the batch of pages/blocks/etc. that make up one
 * logical change lands atomically. That transaction is scoped to `flush()`
 * itself, not to the request as a whole — an earlier version started it lazily
 * on the first read and held it open for the rest of the request, which meant
 * it was still open across `extractDocument()`'s OCR/LLM calls. Those can run
 * well past MongoDB's ~60s transaction lifetime limit, so the server would
 * abort the transaction out from under a still-running request and every
 * write after that point failed with `NoSuchTransaction`. Reads don't need
 * transactional isolation here (nothing reads back an uncommitted write from
 * earlier in the same request), so they run as plain, session-less finds.
 */
export class UnitOfWork {
  readonly db: Db;
  private readonly client: MongoClient;
  private identity = new Map<string, Entity & { __collection: string }>();
  private snapshots = new Map<string, Record<string, unknown> | undefined>();
  private deleted = new Map<string, { __collection: string; id: string }>();

  constructor(db: Db, client: MongoClient) {
    this.db = db;
    this.client = client;
  }

  private key(collection: string, id: string): string {
    return `${collection}:${id}`;
  }

  /** Register an entity read from Mongo, returning the instance already tracked if
   * there is one — two reads of the same row must not diverge in memory. */
  track<T extends Entity>(collection: string, entity: T): T {
    const key = this.key(collection, entity.id);
    const existing = this.identity.get(key);
    if (existing) return existing as unknown as T;
    const tagged = { ...entity, __collection: collection };
    this.identity.set(key, tagged);
    this.snapshots.set(key, toPlain(entity));
    return tagged as unknown as T;
  }

  add<T extends Entity>(collection: string, entity: T): T {
    const key = this.key(collection, entity.id);
    const tagged = { ...entity, __collection: collection };
    this.identity.set(key, tagged);
    this.snapshots.delete(key); // no snapshot -> treated as an insert on flush
    this.deleted.delete(key);
    return tagged as unknown as T;
  }

  delete(collection: string, id: string): void {
    const key = this.key(collection, id);
    this.identity.delete(key);
    this.snapshots.delete(key);
    this.deleted.set(key, { __collection: collection, id });
  }

  async flush(): Promise<void> {
    const deletes = [...this.deleted.values()];

    const writes: Array<{ key: string; collection: string; toWrite: Record<string, unknown> }> = [];
    for (const [key, entity] of this.identity) {
      const { __collection, ...rest } = entity;
      const document = toPlain(rest as unknown as Entity);
      const snapshot = this.snapshots.get(key);
      if (snapshot && shallowEqual(snapshot, document)) continue;
      if (snapshot) {
        (rest as Entity).updatedAt = new Date();
      }
      writes.push({ key, collection: __collection, toWrite: toPlain(rest as unknown as Entity) });
    }

    if (deletes.length === 0 && writes.length === 0) return;

    const clientSession = this.client.startSession();
    try {
      clientSession.startTransaction();
      for (const { __collection, id } of deletes) {
        await this.db.collection<MongoDoc>(__collection).deleteOne({ _id: id }, { session: clientSession });
      }
      for (const { collection, toWrite } of writes) {
        await this.db
          .collection<MongoDoc>(collection)
          .replaceOne({ _id: toWrite.id as string }, mongoDoc(toWrite), { upsert: true, session: clientSession });
      }
      await clientSession.commitTransaction();
    } catch (err) {
      await clientSession.abortTransaction().catch(() => undefined);
      throw err;
    } finally {
      await clientSession.endSession();
    }

    this.deleted.clear();
    for (const { key, toWrite } of writes) this.snapshots.set(key, toWrite);
  }

  async commit(): Promise<void> {
    await this.flush();
  }

  async rollback(): Promise<void> {
    // Writes already made it to Mongo transactionally inside their own flush()
    // call, so there is nothing left to abort here — this only discards the
    // in-memory identity map so a retried request starts clean.
    this.expireAll();
  }

  expireAll(): void {
    this.identity.clear();
    this.snapshots.clear();
    this.deleted.clear();
  }

  async close(): Promise<void> {
    this.expireAll();
  }
}

interface MongoDoc {
  _id: string;
  [key: string]: unknown;
}

function toPlain(entity: Entity): Record<string, unknown> {
  return JSON.parse(JSON.stringify(entity));
}

function mongoDoc(entity: Record<string, unknown>): Record<string, unknown> {
  const { id, ...rest } = entity;
  return { _id: id, ...rest };
}

function shallowEqual(a: Record<string, unknown>, b: Record<string, unknown>): boolean {
  return JSON.stringify(a) === JSON.stringify(b);
}

/** Request-scoped Unit of Work. Call `.close()` in a `finally` block. */
export async function openSession(): Promise<UnitOfWork> {
  const client = await getClientPromise();
  const db = client.db(settings.mongoDbName);
  return new UnitOfWork(db, client);
}

/** Runs `fn` inside a Unit of Work, committing on success and rolling back on throw. */
export async function withSession<T>(fn: (session: UnitOfWork) => Promise<T>): Promise<T> {
  const session = await openSession();
  try {
    const result = await fn(session);
    await session.commit();
    return result;
  } catch (err) {
    await session.rollback();
    throw err;
  } finally {
    await session.close();
  }
}

export { newId };
